[CodeGen] Add aflag to allow potentially to remove unnecessary code to improve performance.

LLITCHEV opened the pull request:
- Added a new flag to allow to test and evaluate performance, so we can evaluate the removal of inter-pass fusion code in the DecomposeSoftmax iree pass that duplicated a sub and exp operation, causing perf degradation.
- Clened up (removed) some of never used/stale code from the pass. It looks like the functinality has been upstreamed to linalg already, but some of the code has remained in the pass.
- Updated the LIT tests as necessary.

LLITCHEV committed:
Removed unnecessary code to improve performance.

Removed code unnecessary code in the DecomposeSoftmax iree pass that duplicated a sub and exp operation, causing unnecessary perf degradation.
Clened up (removed) bunch of never used code from the pass.

LLITCHEV committed:
Fix the LIT tests for softMax to match the new codegen.

LLITCHEV committed:
Testing for Vulcan/SpirV regression.

LLITCHEV committed:
Fixed a minor indentation issue when reenabled old code.

LLITCHEV committed:
Mote testing for Vulcan/SpirV regression.

LLITCHEV committed:
CI test with removed internal fusion.

LLITCHEV committed:
Reverted some temp testing changes.

LLITCHEV committed:
Merge branch 'main' into lubo-issue-311665987

ScottTodd commented on compiler/src/iree/compiler/Codegen/Common/test/decompose_softmax.mlir:
```diff
@@ -19,25 +19,28 @@ func.func @softmax(%arg0: tensor<2x16x32xf32>) -> tensor<2x16x32xf32> {
-// CHECK:          linalg.yield %[[D9]], %[[D10]] : f32
-// CHECK:        } -> (tensor<2x16x32xf32>, tensor<2x16x32xf32>)
-// CHECK:        return %[[D7]]#1 : tensor<2x16x32xf32>
+// CHECK:        %[[D7:.+]] = linalg.generic {indexing_maps = [#[[MAP]], #[[MAP1]], #[[MAP]]], iterator_types =
```
> There are few failures with softMax and they are related to missing Vulcan Extension. The code with the changes in this PR causes the pattern matchers to match for VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_COOPERATIVE_MATRIX_FEATURES_KHR extension (thanks @ScottTodd for pointing where the definition for 1000506000!). 
> ```
> exec ${PAGER:-/usr/bin/less} "$0" || exit 1
> Executing tests from //tests/e2e/regression:check_regression_vulkan-spirv_softmax.mlir
> -----------------------------------------------------------------------------
> [==========] Running 2 tests from 1 test suite.
> [----------] Global test environment set-up.
> [----------] 2 tests from module
> [ RUN ] module.softmax
> /tmp/swiftshader_src_U4O4VS/src/Vulkan/VkPhysicalDevice.cpp:649 WARNING: UNSUPPORTED: curExtension->sType: 1000506000
> 
> The error 1000506000 is
> // Provided by VK_KHR_cooperative_matrix
>     VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_COOPERATIVE_MATRIX_FEATURES_KHR = 1000506000,
> ```
> I'm sure there had been similar problems with this kind of errors and I wonder is someone here knows of examples/ of ways how to deal with these.
> For example there should be probably a way to require for a test to ave certain extensions available for it to run. With that said, I think that the test should be able to run even with no extension available too, maybe?
> How are these type of problems handled? Any examples/patterns used in such cases?

We have labels like `vulkan_uses_vk_khr_shader_float16_int8` and `requires-gpu-rdna3` that can be applied to tests. Each CI build can then filter which tests to include/exclude using those labels.

I'm not sure how we have cooperative matrix wired up right now, but I'd expect it to be optional and enabled based on either the known-at-compile-time Vulkan target or the supported-at-runtime Vulkan device. I don't think for a high level op test we should limit to only devices with a certain extension. It's a bit spooky that this PR can lead to that error, and only in one test... maybe the lowerings for softmax (or some of the included ops) down to SPIR-V are brittle?

qedawkins replied:
cc @kuhar

kuhar replied:
> I'm not sure how we have cooperative matrix wired up right now, but I'd expect it to be optional and enabled based on either the known-at-compile-time Vulkan target or the supported-at-runtime Vulkan device. I don't think for a high level op test we should limit to only devices with a certain extension

Yes, this should be the case. What's the exact compilation command used by this test?

ScottTodd replied:
(I just commented here for threading) The test that failed due to requiring cooperative matrix is `//tests/e2e/regression:check_regression_vulkan-spirv_softmax.mlir` in from https://github.com/openxla/iree/blob/main/tests/e2e/regression/CMakeLists.txt. Compile command is probably `iree-compile --iree-input-type=stablehlo --iree-hal-target-backends=vulkan-spirv tests/e2e/regression/softmax.mlir`

kuhar replied:
Yeah this looks like something that should *not* be doing coop matrix.

ScottTodd reviewed (commented):
I see a bunch of different discussion threads going on here. The code changes are pretty small, so splitting into a smaller PR doesn't seem feasible. How else might we untangle some of these threads?
* Higher bandwidth discussion? (video call)
* More benchmarks / profiles to ground the discussions?
* More test coverage to add better signal for codegen changes?

LLITCHEV committed:
Merge branch 'main' into lubo-issue-311665987

LLITCHEV committed:
Added a flag to control the local fussion as discussed at #20598.
Removed a bunch of dead code and created ability to do furter tsting and potential removal of local fusion for the DecomposeSoftmaxPass, which will potentially result in significant GPU and CPU perf.

LLITCHEV committed:
Merge branch 'main' into lubo-issue-311665987

LLITCHEV committed:
Updated the CMakeList.txt to the test.

Updated the list of test files to reflect the new tests added for the new internal fusion flag.

LLITCHEV committed:
Fix some formatting issues.

LLITCHEV committed:
Added the new test files to blaze builds.

LLITCHEV committed:
More blaze test file conversion changes (order).

LLITCHEV committed:
And more order fixes for the bazel_to_cmake conversion of the test builds.

MaheshRavishankar commented on compiler/src/iree/compiler/Codegen/Common/test/decompose_softmax_local_fusion_false.mlir:
```diff
@@ -0,0 +1,46 @@
+// RUN: iree-opt --split-input-file --iree-codegen-decompose-softmax=use-fusion=false -cse %s | FileCheck %s
```
Instead of separate files, you can just check the different flag values with `--check_prefix=<LABEL>` and combine everything into a single file.

LLITCHEV replied:
Done
Thanks!

MaheshRavishankar reviewed (changes requested):
Thanks! Just a minor nit after which it is good to go. Thanks for also removing the unnecessary code.

LLITCHEV committed:
Used the check-prefix instead of different files for the tests

As per a CR recommendatioj.

LLITCHEV committed:
Removed left over line

LLITCHEV committed:
Changed the basel_to_cmake and reverse test file lists.

LLITCHEV committed:
More test file conversion bazel to cmake.

hanhanW commented on compiler/src/iree/compiler/Codegen/Common/test/decompose_softmax.mlir:
```diff
@@ -1,43 +1,84 @@
-// RUN: iree-opt --split-input-file --iree-codegen-decompose-softmax -cse %s | FileCheck %s
+// RUN: iree-opt --split-input-file --iree-codegen-decompose-softmax -cse %s | FileCheck %s --check-prefix=CHECK-FUSE
+// RUN: iree-opt --split-input-file --iree-codegen-decompose-softmax=use-fusion=false -cse %s | FileCheck %s --check-prefix=CHECK-FUSE
+// RUN: iree-opt --split-input-file --iree-codegen-decompose-softmax=use-fusion=true -cse %s | FileCheck %s --check-prefix=CHECK-NO-FUSE
```
l.1 and l.3 are the same, can we keep the first one? Also, lets not use `check-prefix` for default?

LLITCHEV replied:
@hanhanW 
Removed the check-prefix for the current default case.
The intent of I.1 is to test the default value of the flag and I.3 is testing the behavior with the flag set to false. Just making sure the default for the flag was not changed unintentionally.
Let me know if that makes sense or you prefer it being removed. Thanks!

LLITCHEV committed:
Removed check-prefix for the default case per CR feedback.

MaheshRavishankar reviewed (approved):
Thanks!
